_posts/2018-07-10-Play Framework-in-Scala–Action-Composition.md

```diff
@@ -0,0 +1,256 @@
+- The syntax trough which I create an action is: ThisCustomActionBuilder.apply(block), which, given the syntactic sugar surrounding the apply method in Scala, is the same as ActionBuilder(block) or ActionBuilder { request: R[A] => Result }.
+
+<h4> The syntax for creating default Actions </h4>
+
+When you don't need to define a custom ActionBuilder, Play allows you to write Actions with a very intuitive syntax.
+```scala
+def index(): Action[AnyContent] = Action { request =>
+  Ok("`Action` is actually an ActionBuilder with no common functionality.")
+}
+```
+
+But Action on line 1 is actually a call to a function with a suggestive name (def Action), which returns a default ActionBuilder[Request, AnyContent] which commes along with Play.
+Its invokeBlock is just:
+
+```scala
+def invokeBlock[A](request: Request[A],
+                   block: (Request[A]) => Future[Result]) = block(request)
+```
+
+Because the invokeBlock consists of an immediate call to block, this ActionBuilder describes <strong>no</strong> common functionality for the Actions created through it. It is just a helper for when you want to create an Action straight away.
+
+<h2 id="chaining_actionbuilders"> Chaining ActionBuilders </h2>
+
+So, we have a LoggingActionBuilder that logs the requests, and we have a AuthenticationActionBuilder that authenticates the requests.
+
+What if I want to build Actions whose requests I want to both log an authenticate?
+
+<strong>1. Do I need to create a third ActionBuilder that does both?</strong>
+This has two downsides. Firstly it duplicates logic. If the way we authenticate a request changes, we would have to remember to change on both ActionBuilders.
+Secondly, we would violate the principle of single responsibility, by having the third ActionBuilder both logging and authenticating.
+<strong>2. Can I join ActionBuilders ad-hoc? </strong>
+This is what we want. It would mean we develop custom ActionBuilders, which can be used separately and independently, but that could also be combined together at will. 
+
+Jumping to the result, it should be something like the following sketch:
+
+<img class=" size-full wp-image-398 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/06/chainingactionbuilders.png" alt="ChainingActionBuilders" width="2124" height="494" />
+
+From my contact with Actions, I get the impression Play's Actions were not planned for composition. The documentation is lacking and there is not really many developers discussing it on StackOverflow.
+Still, the problem is underpinned by method <code>andThen</code> of the ActionBuilder.
+
+```scala
+trait ActionBuilder[+R[_], B] extends ActionFunction[Request, R] {
+  ......
+  override def andThen[Q[_]](other: ActionFunction[R, Q]): ActionBuilder[Q, B] = new ActionBuilder[Q, B] {
+    ....
+    def invokeBlock[A](request: Request[A], block: Q[A] => Future[Result]) =
+      self.invokeBlock[A](request, other.invokeBlock[A](_, block))
+  }
+}
+```
```

Comment:
I'd say they were design for comparability, probably the design is not good though.